lib/ui/progress.coffee

```diff
@@ -0,0 +1,30 @@
+{formatTimePeriod} = require '../misc'
+
+module.exports =
+  progs: []
+
+  add: (p) ->
+    pr = @ink.progress.create p
+    pr.register()
+    pr.t0 = Date.now()
+    @progs.push pr
+
+  update: (p) ->
+    pr = @progs.find (pr) => pr.id is p.id
+    pr.setProgress p.progress
+    pr.setLeftText p.leftText
+    if p.rightText?.length
+      pr.setRightText p.rightText
+    else
+      pr.setRightText formatTimePeriod (Date.now() - pr.t0)*(1/p.progress - 1)/1000
```

Comment:
It might be nice to formalise what happens in the right text more at some point. But this is the right approach for now, it lets people provide their own time estimates and stuff at least.